Initial docker-sync support

citizen428 opened the pull request:
## What type of PR is this? (check all applicable)

- [X] Optimization
- [X] Developer experience

## Description

Docker can be very slow on Macs (and Windows) and [docker-sync]() tries to fix that.

## Related Tickets & Documents

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

## Added tests?

- [ ] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [ ] no documentation needed

## What gif best describes this PR or how it makes you feel?

![faster](https://pics.me.me/must-go-faster-gotta-go-fast-25606633.png)

citizen428 commented on docker-compose-sync.yml:
```diff
@@ -0,0 +1,13 @@
+version: "3"
```
This whole file is  an overlay, which `docker-sync` puts "on top of" our normal `docker-compose.yml`. So `docker-sync-stack start` is essentially

```
docker-compose -f docker-compose.yml -f docker-compose-sync.yml up
```

citizen428 commented on docker-entrypoint.sh:
```diff
@@ -14,6 +14,13 @@ echo "alias devto-start='cd /usr/src/app/ && bundle exec rails server -b 0.0.0.0
 cd /usr/src/app/
 
+#
+# Clean out server.pid
```
Technically not necessary for this PR, but useful when someone runs a Docker build while having a server running.

citizen428 commented on docker-sync.yml:
```diff
@@ -0,0 +1,19 @@
+syncs:
+  app-sync:
+    src: './'
+    sync_strategy: 'rsync'
```
Haven't entirely settled on a sync strategy yet, but `rsync` worked better for me than `native_osx`.

maestromac commented:
I receive the following error when I try to run `docker-sync-stack start`

```
error  (<unknown>): did not find expected key while parsing a block mapping at line 7 column 5
```

citizen428 commented:
@maestromac That's slightly confusing, as you shouldn't be using Unison at all, since the sync strategy is specified as `rsync`.

rhymes commented:
This we closed the wrong thing with #6198

citizen428 commented:
@maestromac The error you had was caused by a misconfiguration. I had made some last-minute changes and committed without running again 🤦‍♂ Sorry for that, I fixed up the config now and added some docs. If you could try it out and tell me if/where it goes wrong that would be super helpful! Thanks!

maestromac commented:
@citizen428 I was able to start it up! This time I see the following error for the web section
```
eb_1            | bundler: failed to load command: rails (/usr/local/bundle/bin/rails)
web_1            | Bundler::GemNotFound: Could not find concurrent-ruby-1.1.6 in any of the sources
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler/spec_set.rb:87:in `block in materialize'
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler/spec_set.rb:81:in `map!'
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler/spec_set.rb:81:in `materialize'
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler/definition.rb:170:in `specs'
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler/definition.rb:237:in `specs_for'
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler/definition.rb:226:in `requested_specs'
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler/runtime.rb:108:in `block in definition_method'
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler/runtime.rb:20:in `setup'
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler.rb:107:in `setup'
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler/setup.rb:20:in `<top (required)>'
web_1            |   /usr/local/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
web_1            |   /usr/local/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
elasticsearch    | OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
devto_web_1 exited with code 1
```

citizen428 commented:
Hm, @maestromac, can you try running `docker-compose build` before following the rest of the documentation?

maestromac commented:
I've been unsuccessful to docker up and running, particularly on the `docker-compose run web rails search:setup` step I get the following.
```
Elasticsearch::Transport::Transport::Errors::BadRequest: [400] {"error":{"root_cause":[{"type":"invalid_alias_name_exception","reason":"Invalid alias name [tags_development_alias], an index exists with the same name as the alias","index_uuid":"DriMXz20SnWTvC2PZC9mlw","index":"tags_development_alias"}],"type":"invalid_alias_name_exception","reason":"Invalid alias name [tags_development_alias], an index exists with the same name as the alias","index_uuid":"DriMXz20SnWTvC2PZC9mlw","index":"tags_development_alias"},"status":400}
```

 When I do try to start docker-sync though I get a Airbrake (which was removed) related error.

```
 /usr/local/bundle/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/active_support.rb:79:in `block in load_missing_constant': uninitialized constant Airbrake (NameError)
```

This could just be the consequence of constantly switching the branch back and forth. I believe docker-sync is working now because another instance of the codebase(pg and redis) is running even though the web failed. Are you able to get it working 100% on your machine?

citizen428 commented:
Yes, it seems to be working just fine for me. Maybe it would help if you clean up your Docker environment a bit (if you're not using it for anything else `docker system prune` is a fast way to do that).

citizen428 commented:
Yes, it seems to be working just fine for me. Maybe it would help if you clean up your Docker environment a bit (if you're not using it for anything else `docker system prune` is a fast way to do that).

maestromac commented:
@citizen428 I gave these multiple attempts and I always run into this issue.

```
web_1            | bundler: failed to load command: rails (/usr/local/bundle/bin/rails)
web_1            | Bundler::GemNotFound: Could not find nokogiri-1.10.7 in any of the sources
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler/spec_set.rb:87:in `block in materialize'
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler/spec_set.rb:81:in `map!'
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler/spec_set.rb:81:in `materialize'
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler/definition.rb:170:in `specs'
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler/definition.rb:237:in `specs_for'
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler/definition.rb:226:in `requested_specs'
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler/runtime.rb:108:in `block in definition_method'
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler/runtime.rb:20:in `setup'
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler.rb:107:in `setup'
web_1            |   /usr/local/bundle/gems/bundler-2.0.2/lib/bundler/setup.rb:20:in `<top (required)>'
web_1            |   /usr/local/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
web_1            |   /usr/local/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
elasticsearch    | OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
devto_web_1 exited with code 1
```

Maybe this is caused by the config?

citizen428 commented:
@maestromac Just to make sure we are not confusing issues, do you only get this when trying to run docker-sync, or also with the normal docker-compose workflow?

maestromac commented:
I believe it was docker-sync specific issue but I need to sort out my personal docker setup first. There's a different issue every single time I try to start it up and it takes me over 30 min to clean and install every single time

maestromac commented on docs/installation/docker.md:
```diff
@@ -86,6 +86,18 @@ Please execute the script itself to view all additional options:
+following steps:
+
+1. Install `docker-sync`: `gem install docker-sync`
+2. Make sure your system has `rsync` available
```
I believe you are missing a `docker-compose build` step here

citizen428 replied:
I do indeed. Funny, it definitely was there at one point. Sorry for that!

mstruve commented:
@citizen428 do we still need this?

citizen428 commented:
@mstruve In the sense that Docker is still slow on Macs, yes. But since no one except for @maestromac engaged with this PR, I guess we don't have enough Mac users who feel the pain, so we can just close it.
